Digest consolidation

Papierkorb opened the pull request:
Hello,

Here's the Digest consolidation as promised in #4021 :)

The original MD5 implementation used a public `Context` class and a private `ContextImpl` struct. I guess the original creators intention was to save a memory allocation in the non-yielding digest methods, and use memory allocation only for the yielding digest methods.

After a discussion in IRC, we came to the conclusion to further improve upon it by make it not use any dynamic memory allocation. As this is not the first time I needed something like this (And likely, others did too), we agreed to put a generic wrapper into the stdlib, called `ValueRef(T)`. This wrapper basically mimics a C++ variable reference.

I'm eager to hear y'all opinions, if this is ok, or if the `ValueRef(T)` should not be there at all, or be part of its own PR :)

(Side note: with that there are now 5000 stdlib unit tests. Yay)

ysbaddaden commented on src/digest/base.cr:
```diff
@@ -0,0 +1,83 @@
+require "value_ref"
+require "base64"
+
+module Digest::Base(C)
```
Please use `T` for denoting a generic.

RX14 replied:
Why? Naming standards we all agree on (i.e. dont use 1 letter variable names) seem to get routinely ignored when using generics. Wouldn't `ContextType` be a better name? `T`  is a good generic name for generic containers, this is not a generic container.

ysbaddaden replied:
Convention. Reading `C` made we wondered wth it was, and why someone would call a struct/class like that, until I noticed it denoted a generic. Reading `T` I immediately understand it as a generic type.

RX14 replied:
I think it's an awful shame to add the feature so that our generic variables can have descriptive names, but not use it. Non-generic type names tend not to end in `Type` for obvious reasons, but if we made it the convention that multi-char generic names ended in `Type` would that make it more readable for you?

Papierkorb commented:
> Maybe dropping the block form would be simpler?

The block form allows one to calculate a hash of large data easily, without having everything in memory. Use-case would be hashing large files.

asterite commented:
I'm not entirely sure about the current design of digest. Providing only the block form is not good, we should provide a way to create a digest, maybe store it in an instance variable and update it as data comes. For this we need to use a class. But writing `@digest : ValueRef(Digest::MD5::Context)` is not very nice. I think we should just make `Digest::MD5::Context` a class, without having a structure. Later the compiler could remove this extra allocation with escape analysis. But we shouldn't complicate the design of the standard library because of this small allocation.

Papierkorb commented:
> But writing @digest : ValueRef(Digest::MD5::Context) is not very nice.

That would've been inherently unsafe too, `ValueRef` wasn't intended to be stored.

> Later the compiler could remove this extra allocation with escape analysis.

Ah, that sounds really good! In that case, I don't have much qualms about using a `class` there "for now", when the compiler will get rid of the expensive parts "later on" anyway.

> But we shouldn't complicate the design of the standard library because of this small allocation.

Agreed.

Latest changes: Removed `ValueRef` again and renamed the ContextImpl structs to `class Context`

asterite commented:
I'm also not sure we should design this in a PR. For example there's no reason to use a generic module here.

We'll some time in the future design this and get back to this.

Papierkorb commented:
So while we wait for the CI to finish, the current PR contains:

1. Renamed `Crypto::MD5` to `Digest::MD5`
2. Moved the class methods from `MD5` and `SHA1` into `Digest::Base` (`#digest`, `#hexdigest`, `#base64digest`, each with and without **yield**)
3. Both `MD5` and `SHA1` inherit from `Digest::Base`
4. `MD5` and `SHA1` have their logic right in the class, and not in a `ContextImpl` sub-struct
5. The `Context` class has been removed and basically replaced by the classes themself

spalladino commented:
Merged, thanks for the contribution @Papierkorb!
